File: src/main/java/net/minestom/server/utils/collection/MergedMap.java
package net.minestom.server.utils.collection;

import org.jetbrains.annotations.ApiStatus;

import java.util.*;
import java.util.function.BiConsumer;
import java.util.function.Predicate;
import java.util.stream.Stream;

@ApiStatus.Internal
public final class MergedMap<K, V> extends AbstractMap<K, V> {
    private final Map<K, V> first, second;

    public MergedMap(Map<K, V> first, Map<K, V> second) {
        this.first = Objects.requireNonNull(first);
        this.second = Objects.requireNonNull(second);
    }

    // mandatory methods

    final Set<Entry<K, V>> entrySet = new AbstractSet<>() {
        @Override
        public Iterator<Map.Entry<K, V>> iterator() {
            return stream().iterator();
        }

        @Override
        public int size() {
            return (int) stream().count();
        }

        @Override
        public Stream<Entry<K, V>> stream() {
            return Stream.concat(first.entrySet().stream(), secondStream())
                    .map(e -> new AbstractMap.SimpleImmutableEntry<>(e.getKey(), e.getValue()));
        }

        @Override
        public Stream<Entry<K, V>> parallelStream() {
            return stream().parallel();
        }

        @Override
        public Spliterator<Entry<K, V>> spliterator() {
            return stream().spliterator();
        }
    };

    Stream<Entry<K, V>> secondStream() {
        return second.entrySet().stream().filter(e -> !first.containsKey(e.getKey()));
    }

    @Override
    public Set<Map.Entry<K, V>> entrySet() {
        return entrySet;
    }

    // optimizations

    @Override
    public boolean containsKey(Object key) {
        return first.containsKey(key) || second.containsKey(key);
    }

    @Override
    public boolean containsValue(Object value) {
        return first.containsValue(value) ||
                secondStream().anyMatch(Predicate.isEqual(value));
    }

    @Override
    public V get(Object key) {
        V v = first.get(key);
        return v != null ? v : second.get(key);
    }

    @Override
    public V getOrDefault(Object key, V defaultValue) {
        V v = first.get(key);
        return v != null ? v : second.getOrDefault(key, defaultValue);
    }

    @Override
    public void forEach(BiConsumer<? super K, ? super V> action) {
        first.forEach(action);
        second.forEach((k, v) -> {
            if (!first.containsKey(k)) action.accept(k, v);
        });
    }
}

File: src/main/java/net/minestom/server/utils/collection/MergedMap.kt
package net.minestom.server.utils.collection

import java.util.AbstractMap.SimpleImmutableEntry
import java.util.function.BiConsumer
import java.util.function.IntFunction
import java.lang.IndexOutOfBoundsException
import java.util.concurrent.atomic.AtomicReference
import net.minestom.server.utils.collection.MappedCollection
import org.jetbrains.annotations.ApiStatus
import java.util.*
import java.util.function.Predicate
import java.util.stream.Stream

@ApiStatus.Internal
class MergedMap<K, V>(first: Map<K, V>, second: Map<K, V>) : AbstractMap<K?, V?>() {
    private val first: Map<K, V>
    private val second: Map<K, V>

    // mandatory methods
    val entrySet: Set<Map.Entry<K, V>> = object : AbstractSet<Map.Entry<K, V>?>() {
        override fun iterator(): MutableIterator<Map.Entry<K, V>> {
            return stream().iterator()
        }

        override fun size(): Int {
            return stream().count().toInt()
        }

        override fun stream(): Stream<Map.Entry<K, V>?> {
            return Stream.concat(first.entries.stream(), secondStream())
                .map { (key, value): Map.Entry<K, V> ->
                    SimpleImmutableEntry(
                        key, value
                    )
                }
        }

        override fun parallelStream(): Stream<Map.Entry<K, V>?> {
            return stream().parallel()
        }

        override fun spliterator(): Spliterator<Map.Entry<K, V>?> {
            return stream().spliterator()
        }
    }

    init {
        this.first = Objects.requireNonNull(first)
        this.second = Objects.requireNonNull(second)
    }

    fun secondStream(): Stream<Map.Entry<K, V>> {
        return second.entries.stream().filter { (key): Map.Entry<K, V> -> !first.containsKey(key) }
    }

    override fun entrySet(): Set<Map.Entry<K, V>> {
        return entrySet
    }

    // optimizations
    override fun containsKey(key: Any?): Boolean {
        return first.containsKey(key) || second.containsKey(key)
    }

    override fun containsValue(value: Any?): Boolean {
        return first.containsValue(value) ||
                secondStream().anyMatch(Predicate.isEqual(value))
    }

    override operator fun get(key: Any?): V? {
        val v = first[key]
        return v ?: second[key]
    }

    override fun getOrDefault(key: Any, defaultValue: V): V? {
        val v = first.get(key)
        return v ?: second.getOrDefault(key, defaultValue)
    }

    override fun forEach(action: BiConsumer<in K, in V>) {
        first.forEach(action)
        second.forEach { (k: K, v: V) -> if (!first.containsKey(k)) action.accept(k, v) }
    }
}